### ISSUE TITLE
Streaming accuracy and recall aren't working as expected

### ISSUE THREAD

**Issue by ihsangunay** [tensorflow/tensorflow#8375]
### Environment info
Ubuntu 16.04
Cuda 8.0
Cudnn 5.1

1. The commit hash (`git rev-parse HEAD`)
c56c873fbaf976d26d487ad57c8efbc87f05331c
2. The output of `bazel version`
Build label: 0.4.4
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Feb 1 18:54:21 2017 (1485975261)
Build timestamp: 1485975261
Build timestamp as int: 1485975261

Here is my code:
```
weights = {'first': tf.Variable(tf.random_normal([1, 3, 1, 10])),
           'iterated': tf.Variable(tf.random_normal([1, 3, 10, 10])),
           'out': tf.Variable(tf.random_normal([embedding_dim*10, n_classes]))}

biases = {'first': tf.Variable(tf.random_normal([10])),
          'iterated': tf.Variable(tf.random_normal([10])),
          'out': tf.Variable(tf.random_normal([n_classes]))}

preds, cost = model(x, y, weights, biases, dropout, depth_tensor)

optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(cost)

accuracy, update_accuracy = streaming_accuracy(y, preds)
recall, update_recall = streaming_recall(y, preds)

init = tf.global_variables_initializer()
init2 = tf.local_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    i = 1
    
    for batch_x, batch_y in data_processor(data, train_inds, embedding, label_processor, n_iter):
        sess.run(optimizer, 
                 feed_dict={x: batch_x, y: batch_y, 
                            dropout: dropout_prob})
        
        if i % display_step == 0:
            loss = sess.run(cost, 
                            feed_dict={x: batch_x, y: batch_y, dropout: dropout_prob})
            
            print("Iter:{}, Minibatch Loss:{:.6f}".format(i,loss))
        i += 1
    
    sess.run(init2)
    for batch_x, batch_y in data_processor(data, val_inds, embedding, label_processor, n_iter):
        recall, accuracy = sess.run([update_recall, update_accuracy], 
                                    feed_dict={x:batch_x, y: batch_y, dropout: 1})
        
        f1 = 2 * recall * accuracy / (recall + accuracy)
    
    print("Testing Accuracy:", accuracy,"Testing Recall:", recall, "Testing F1 Score:", f1) 
```

Output:
```
Iter:100, Minibatch Loss:18038.144531
Iter:200, Minibatch Loss:11628.046875
Iter:300, Minibatch Loss:9288.974609
Iter:400, Minibatch Loss:4583.474121
Iter:500, Minibatch Loss:6600.524902
...
Iter:11700, Minibatch Loss:4.203137
Iter:11800, Minibatch Loss:3.623320
Iter:11900, Minibatch Loss:4.883300
Iter:12000, Minibatch Loss:3.045975
Testing Accuracy: 0.0 Testing Recall: 0.00211863 Testing F1 Score: 0.0
```

**Comment by drpngx** [comment]
It looks like this is more of a usage question. The best venue is to ask on stackoverflow, where we monitor all questions with the tag `tensorflow`. From the output, it looks like your model is still not working well. You should probably continue to train, and perhaps make your model smaller.

Closing here, feel free to post a cross-link to stack overflow.

**Comment by randomrandom** [comment]
I also get strangely small recall for multi-class problem. Here's a comparison between the metrics I get via scikit-learn and via Tensorflow:
**Scikit-learn code**
```
        all_true = []
        all_predicted = []
        for i in tqdm(range(0, EPOCHS * test.data_size // BATCH_SIZE)):
            entities_sample, predictions_sample, _, __ = sess.run([test.entities, predictions, precision_op, recall_op])

            all_true.extend(entities_sample.flatten())
            all_predicted.extend(predictions_sample.flatten())

        s_prec = metrics.precision_score(all_true, all_predicted, labels=[1, 2, 3, 4, 5, 6, 7, 8, 9], average='micro')
        s_rec = metrics.recall_score(all_true, all_predicted, labels=[1, 2, 3, 4, 5, 6, 7, 8, 9], average='micro')
        s_f1 = metrics.f1_score(all_true, all_predicted, labels=[1, 2, 3, 4, 5, 6, 7, 8, 9], average='micro')
        s_confusion = metrics.confusion_matrix(all_true, all_predicted)

        print(s_prec)
        print(s_rec)
        print(s_f1)
```

**Scikit-learn output**

```
0.875761847506
0.875743692863
0.87575277009
```

**Tensorflow code**

```
        for i in tqdm(range(0, EPOCHS * test.data_size // BATCH_SIZE)):
            entities_sample, predictions_sample, _, __ = sess.run([test.entities, predictions, precision_op, recall_op])

            all_true.extend(entities_sample.flatten())
            all_predicted.extend(predictions_sample.flatten())


        f1_score = (2 * (precision * recall)) / (precision + recall)

        final_precision, final_recall, final_f1 = sess.run([precision, recall, f1_score])
        print('Precision:{}'.format(final_precision))
        print('Recall:{}'.format(final_recall))
        print('f-1 score:{}'.format(final_f1))
```

**Tensorflow output**

```
Precision:0.8595223781587512
Recall:0.4297611890793756
f-1 score:0.5730149187725008
```

As you can see the difference between scikit's recall and tensorflow is quite drastic. Not sure if this is a bug or a difference in the implementation. I've tried calculating the recall with scikit with other various options, described here: [http://scikit-learn.org/stable/modules/model_evaluation.html#precision-recall-and-f-measures]() but never achieved such low results

**Comment by muayyad-alsadi** [comment]
From what I see, tensorflow's recall and precision do not mean what they are supposed to mean

modified from this example from [here](https://github.com/tensorflow/models/blob/master/research/slim/eval_image_classifier.py)

```
    predictions = tf.argmax(logits, 1)
    labels = tf.squeeze(labels)
    names_to_values, names_to_updates = slim.metrics.aggregate_metric_map({
        'Accuracy': slim.metrics.streaming_accuracy(predictions, labels),
        'Precision': slim.metrics.streaming_precision(predictions, labels),
        'Recall': slim.metrics.streaming_recall(predictions, labels),
        'Recall_5': slim.metrics.streaming_recall_at_k(logits, labels, 5),
        'Recall_3': slim.metrics.streaming_recall_at_k(logits, labels, 3),
        'Recall_1': slim.metrics.streaming_recall_at_k(logits, labels, 1),
        })
```

```
2018-03-06 12:45:43.520961: I tensorflow/core/kernels/logging_ops.cc:79] eval/Recall_1[0.664843738]
2018-03-06 12:45:43.521368: I tensorflow/core/kernels/logging_ops.cc:79] eval/Recall[0.990521312]
2018-03-06 12:45:43.521429: I tensorflow/core/kernels/logging_ops.cc:79] eval/Recall_5[0.857031226]
2018-03-06 12:45:43.521487: I tensorflow/core/kernels/logging_ops.cc:79] eval/Precision[0.996820331]
2018-03-06 12:45:43.521537: I tensorflow/core/kernels/logging_ops.cc:79] eval/Accuracy[0.664843738]
2018-03-06 12:45:43.521584: I tensorflow/core/kernels/logging_ops.cc:79] eval/Recall_3[0.809375]
```

how come both streaming_recall and streaming_precision are 99% while accuracy and top 1 recall are 66%.

something seriously is different than the known meaning of [recall and precision](https://en.wikipedia.org/wiki/Precision_and_recall)
we know.  and why accuracy is same as recall_1 and why recall and recall_1 are different?

**Comment by muayyad-alsadi** [comment]
here is the stackoverflow link

https://stackoverflow.com/questions/49131436/tensorflows-recall-and-precision-do-not-mean-what-they-are-supposed-to-mean

**Comment by muayyad-alsadi** [comment]
related issue

https://stackoverflow.com/questions/43618608/tensorflow-streaming-metrics-are-not-working?rq=1
